community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/EncodeGroupsStep.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.unsafe.impl.batchimport;
+
+import org.neo4j.kernel.impl.store.RecordStore;
+import org.neo4j.kernel.impl.store.record.Record;
+import org.neo4j.kernel.impl.store.record.RelationshipGroupRecord;
+import org.neo4j.unsafe.impl.batchimport.staging.BatchSender;
+import org.neo4j.unsafe.impl.batchimport.staging.Configuration;
+import org.neo4j.unsafe.impl.batchimport.staging.ProcessorStep;
+import org.neo4j.unsafe.impl.batchimport.staging.StageControl;
+
+/**
+ * Takes cached {@link RelationshipGroupRecord relationship groups} and sets real ids and
+ * {@link RelationshipGroupRecord#getNext() next pointers}, making them ready for writing to store.
+ */
+public class EncodeGroupsStep extends ProcessorStep<RelationshipGroupRecord[]>
+{
+    private long nextId = -1;
+    private final RecordStore<RelationshipGroupRecord> store;
+
+    public EncodeGroupsStep( StageControl control, Configuration config, RecordStore<RelationshipGroupRecord> store )
+    {
+        super( control, "ENCODE", config, 1 );
+        this.store = store;
+    }
+
+    @Override
+    protected void process( RelationshipGroupRecord[] batch, BatchSender sender ) throws Throwable
+    {
+        for ( int i = 0; i < batch.length; i++ )
+        {
+            RelationshipGroupRecord group = batch[i];
```

Comment:
would it make more sense here to do processing of groups in a first loop - so you can avoid dance with next id and nextid + 10 thingy
and then as a second loop just do prepareForCommit for all of them?